### PR TITLE
Mac sleep monitor

### PR DESCRIPTION
**Pull request by howard-stearns**
OSX addition for https://highfidelity.atlassian.net/browse/BUGZ-946 / https://github.com/highfidelity/hifi/pull/15898

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20352/HighFidelity-Beta-PR15919-cc7214f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20352/HighFidelity-Beta-Interface-PR15919-cc7214f.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20352/HighFidelity-Beta-PR15919-cc7214f.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20352/HighFidelity-Quest-Beta-PR15919-cc7214f.apk)

**Comment by howard-stearns** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20357/HighFidelity-Beta-PR15919-cc7214f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20357/HighFidelity-Beta-Interface-PR15919-cc7214f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20357/HighFidelity-Beta-PR15919-cc7214f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20357/HighFidelity-Beta-Interface-PR15919-cc7214f.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20357/HighFidelity-Beta-PR15919-cc7214f.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20357/HighFidelity-Quest-Beta-PR15919-cc7214f.apk)

File: interface/src/MacHelper.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#include <NodeList.h>
+
+#ifdef Q_OS_MAC
+#include <IOKit/IOMessage.h>
+#include <IOKit/pwr_mgt/IOPMLib.h>
+
+// These type definitions come from IOKit, which includes a definition of Duration that conflicts with ours.
+// So... we include these definitions here rather than in the .h, as the .h is included in Application.cpp which
+// uses Duration.
+static io_connect_t  root_port;
```
Comment (SimonWalton-HiFi) [inline interface/src/MacHelper.cpp]: These are actually object definitions, not type definitions, right? Couldn't they just be members of MacHelper?

Comment (howard-stearns) [reply interface/src/MacHelper.cpp]: Not if MacHelper.h is included in Application.cpp.
There are ways... and I'm open to suggestions. But my instinct is always to go with good old straight C.

File: interface/src/MacHelper.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+static IONotificationPortRef notifyPortRef;
+static io_object_t notifierObject;
+static void* refCon;
+
+static void sleepHandler(void* refCon, io_service_t service, natural_t messageType, void* messageArgument) {
+    if (messageType == kIOMessageSystemHasPoweredOn) {
+        qCInfo(interfaceapp) << "Waking up from sleep or hybernation.";
+        QMetaObject::invokeMethod(DependencyManager::get<NodeList>().data(), "noteAwakening", Qt::QueuedConnection);
```
Comment (roxanneskelly) [inline interface/src/MacHelper.cpp]: I think we were talking about adding 'Awakening' or something to the Connect Reason in the server logs at some point.  I assume this is coming and will be triggered off of this method invocation.

Comment (howard-stearns) [reply interface/src/MacHelper.cpp]: That's in #15898, which is already merged. This PR is just the mac follow-up to that.

File: interface/src/MacHelper.cpp (hunk 1 above)
Comment (howard-stearns) [reply interface/src/MacHelper.cpp]: (the mumble_t are typedefs)

Comment (SimonWalton-HiFi) [reply interface/src/MacHelper.cpp]: I get that, I was just reading your comment as saying these lines were defining types, not variables.

Comment (howard-stearns) [reply interface/src/MacHelper.cpp]: got it. How about I change it to: "The type definitions in these variables come from IOKit..." ?

Comment (SimonWalton-HiFi) [reply interface/src/MacHelper.cpp]: That would be awesome!

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20361/HighFidelity-Beta-PR15919-7ce6dd7.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20361/HighFidelity-Beta-Interface-PR15919-7ce6dd7.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20361/HighFidelity-Beta-PR15919-7ce6dd7.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20361/HighFidelity-Beta-Interface-PR15919-7ce6dd7.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20361/HighFidelity-Beta-PR15919-7ce6dd7.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20361/HighFidelity-Quest-Beta-PR15919-7ce6dd7.apk)

**Pull request merged**